daemon/algod/api/server/v2/handlers.go

```diff
@@ -1192,6 +1194,90 @@ func (v2 *Handlers) GetApplicationByID(ctx echo.Context, applicationID uint64) e
 	return ctx.JSON(http.StatusOK, response)
 }
 
+func applicationBoxesMaxKeys(requestedMax uint64, algodMax uint64) uint64 {
+	if requestedMax == 0 {
+		if algodMax == 0 {
+			return 0 // unlimited results when both requested and algod max are 0
+		}
+		return algodMax + 1 // API limit dominates.  Increments by 1 to test if more than max supported results exist.
+	}
+
+	if requestedMax <= algodMax || algodMax == 0 {
+		return requestedMax // requested limit dominates
+	}
+
+	return algodMax + 1 // API limit dominates.  Increments by 1 to test if more than max supported results exist.
+}
+
+// GetApplicationBoxes returns the box names of an application
+// (GET /v2/applications/{application-id}/boxes)
+func (v2 *Handlers) GetApplicationBoxes(ctx echo.Context, applicationID uint64, params generated.GetApplicationBoxesParams) error {
+	appIdx := basics.AppIndex(applicationID)
+	ledger := v2.Node.LedgerForAPI()
+	lastRound := ledger.Latest()
+	keyPrefix := logic.MakeBoxKey(appIdx, "")
+
+	// cast param.Max to zero, if param.Max points to some value, we keep the value
+	castedMax := nilToZero(params.Max)
+	maxBoxThreshold := v2.Node.Config().MaxAPIBoxPerApplication
+	algodLimitsResults := maxBoxThreshold > 0
+
+	boxKeys, err := ledger.LookupKeysByPrefix(lastRound, keyPrefix, applicationBoxesMaxKeys(castedMax, maxBoxThreshold))
+	if err != nil {
+		return internalError(ctx, err, errFailedLookingUpLedger, v2.Log)
+	}
+
+	if algodLimitsResults && uint64(len(boxKeys)) > maxBoxThreshold {
```

Comment:
I don't think it can, except by first going to the db to do a `select count(*) ...`

While that would dave time in the failure case, it would make all the normal cases slower.

Reply:
Re-opening based on the review conversation. The app account has a total so we can avoid looking up boxes if there are going to be too many.

Reply:
@winder Thanks for the added context. 

Re-resolving only because there's multiple threads on the topic.  Let's use https://github.com/algorand/go-algorand/pull/4149#discussion_r1001083810 as the single place to continue discussion here. 

I think https://github.com/algorand/go-algorand/pull/4149#discussion_r1001083810 captures the requirements from this thread.  If you feel something is missing, let's go there for discussion.